ui/src/app/cohort-search/cohort-search/cohort-search.component.tsx

```diff
@@ -0,0 +1,225 @@
+  setShowWarningModal: (showWarningModal: () => Promise<boolean>) => void;
+  workspace: WorkspaceData;
+}
+
+interface State {
+  loading: boolean;
+  overview: boolean;
+  criteria: SearchRequest;
+  updateCount: number;
+  cohort: Cohort;
+  minHeight: string;
+  modalPromise: Promise<boolean> | null;
+  modalOpen: boolean;
+  updateGroupListsCount: number;
+  cohortChanged: boolean;
+  searchContext: any;
+}
+
+export const CohortSearch = withCurrentWorkspace() (
+  class extends React.Component<Props, State> {
+    private subscription;
+    resolve: Function;
+    searchWrapper: HTMLDivElement;
+
+    constructor(props: any) {
+      super(props);
+      this.state = {
+        loading: false,
+        overview: false,
+        criteria: {includes: [], excludes: [], dataFilters: []},
+        updateCount: 0,
+        cohort: undefined,
+        minHeight: '10rem',
+        modalPromise:  null,
+        modalOpen: false,
+        updateGroupListsCount: 0,
+        cohortChanged: false,
+        searchContext: undefined
+      };
+      this.showWarningModal = this.showWarningModal.bind(this);
+    }
+
+    componentDidMount() {
+      const {workspace: {id, namespace}} = this.props;
+      this.subscription = queryParamsStore.subscribe(params => {
+        /* If a cohort id is given in the route, we initialize state with
+         * it */
+        const cohortId = params.cohortId;
+        if (cohortId) {
+          this.setState({loading: true});
+          cohortsApi().getCohort(namespace, id, cohortId).then(cohort => {
+            this.setState({cohort, loading: false});
+            currentCohortStore.next(cohort);
+            if (cohort.criteria) {
+              searchRequestStore.next(parseCohortDefinition(cohort.criteria));
+            }
+          });
+        } else {
+          this.setState({cohort: {criteria: `{'includes':[],'excludes':[],'dataFilters':[]}`, name: '', type: ''}});
+        }
+      });
+
+      this.subscription.add(searchRequestStore.subscribe(sr => {
+        const cohortChanged = !!this.state.cohort && this.state.cohort.criteria !== JSON.stringify(mapRequest(sr));
+        this.props.setCohortChanged(cohortChanged);
+        this.setState({
+          criteria: sr,
+          overview: sr.includes.length > 0 || sr.excludes.length > 0,
+          cohortChanged,
+          updateGroupListsCount: this.state.updateGroupListsCount + 1
+        });
+      }));
+      this.updateWrapperDimensions();
+      this.props.setShowWarningModal(this.showWarningModal);
+    }
+
+    componentWillUnmount() {
+      this.subscription.unsubscribe();
+      idsInUse.next(new Set());
+      currentCohortStore.next(undefined);
+      searchRequestStore.next({includes: [], excludes: [], dataFilters: []} as SearchRequest);
+    }
+
+    async showWarningModal() {
+      this.setState({modalOpen: true});
+      return await new Promise<boolean>((resolve => this.resolve = resolve));
```

Comment:
Question: Whats _await new Promise<boolean>((resolve => this.resolve = resolve));_ used for